frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,382 @@ pub mod pallet {
 			let reasons = WithdrawReasons::TRANSFER | WithdrawReasons::RESERVE;
-			T::Currency::set_lock(VESTING_ID, &who, locked_now, reasons);
-			Self::deposit_event(Event::<T>::VestingUpdated(who, locked_now));
+			T::Currency::set_lock(VESTING_ID, who, total_locked_now, reasons);
+			Self::deposit_event(Event::<T>::VestingUpdated(who.clone(), total_locked_now));
+		};
+	}
+
+	/// Write an accounts updated vesting schedules to storage.
+	fn write_vesting(
+		who: &T::AccountId,
+		schedules: Vec<VestingInfo<BalanceOf<T>, T::BlockNumber>>,
+	) -> Result<(), DispatchError> {
+		let schedules: BoundedVec<
+			VestingInfo<BalanceOf<T>, T::BlockNumber>,
+			T::MaxVestingSchedules,
+		> = schedules.try_into().map_err(|_| Error::<T>::AtMaxVestingSchedules)?;
+
+		if schedules.len() == 0 {
+			Vesting::<T>::remove(&who);
+		} else {
+			Vesting::<T>::insert(who, schedules)
 		}
+
+		Ok(())
+	}
+
+	/// Unlock any vested funds of `who`.
+	fn do_vest(who: T::AccountId) -> DispatchResult {
+		let schedules = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+
+		let (schedules, locked_now) =
+			Self::report_schedule_updates(schedules.to_vec(), VestingAction::Passive);
+
+		debug_assert!(
+			locked_now > Zero::zero() && schedules.len() > 0 ||
+				locked_now == Zero::zero() && schedules.len() == 0
+		);
+		Self::write_vesting(&who, schedules)?;
+		Self::write_lock(&who, locked_now);
+
 		Ok(())
 	}
 }
 
-impl<T: Config> VestingSchedule<T::AccountId> for Pallet<T> where
-	BalanceOf<T>: MaybeSerializeDeserialize + Debug
+impl<T: Config> VestingSchedule<T::AccountId> for Pallet<T>
+where
+	BalanceOf<T>: MaybeSerializeDeserialize + Debug,
 {
-	type Moment = T::BlockNumber;
 	type Currency = T::Currency;
+	type Moment = T::BlockNumber;
 
 	/// Get the amount that is currently being vested and cannot be transferred out of this account.
 	fn vesting_balance(who: &T::AccountId) -> Option<BalanceOf<T>> {
 		if let Some(v) = Self::vesting(who) {
 			let now = <frame_system::Pallet<T>>::block_number();
-			let locked_now = v.locked_at::<T::BlockNumberToBalance>(now);
-			Some(T::Currency::free_balance(who).min(locked_now))
+			let total_locked_now = v.iter().fold(Zero::zero(), |total, schedule| {
+				schedule.locked_at::<T::BlockNumberToBalance>(now).saturating_add(total)
+			});
+			Some(T::Currency::free_balance(who).min(total_locked_now))
 		} else {
 			None
 		}
 	}
 
 	/// Adds a vesting schedule to a given account.
 	///
-	/// If there already exists a vesting schedule for the given account, an `Err` is returned
-	/// and nothing is updated.
+	/// If the account has `MaxVestingSchedules`, an Error is returned and nothing
+	/// is updated.
 	///
 	/// On success, a linearly reducing amount of funds will be locked. In order to realise any
 	/// reduction of the lock over time as it diminishes, the account owner must use `vest` or
 	/// `vest_other`.
 	///
 	/// Is a no-op if the amount to be vested is zero.
+	/// NOTE: does not validate schedule params.
```

Comment:
should check, why not?

Should return error if `per_block` is zero

Reply:
24214ac2e0f049504c29057a88856c9d7ca02da5

I now call `VestingInfo.validate` in here and `can_add_vesting_schedule`. This will fail if: `per_block` or `locked` are 0 or ending block is greater than max block.

Lmk if you think that is too strict tho (i.e. you just want to check for 0)